### ISSUE TITLE
Longer block processing time with 9.51

### ISSUE THREAD

**Issue by mtbitcoin** [CounterpartyXCP/counterpartyd#748]
This is a follow up from the slack conversation regarding the longer block processing times .. 

For example :

[INFO] Block: 350493 (766.42s)←[0m
[INFO] Block: 350494 (770.62s)←[0m

Does this imply that the both blocks above took around 12min +-  . ??

**Comment by deweller** [comment]
I'm seeing similar results from v9.50.0

```
2015-04-03-T11:03:01+0000 [INFO] Block: 350526 (617.27s)
2015-04-03-T11:13:50+0000 [INFO] Block: 350527 (649.19s)
2015-04-03-T11:13:59+0000 [INFO] Block: 350528 (8.83s)
2015-04-03-T11:24:35+0000 [INFO] Block: 350529 (636.33s)
2015-04-03-T11:38:30+0000 [INFO] Block: 350530 (361.19s)
2015-04-03-T11:49:34+0000 [INFO] Block: 350531 (572.60s)
2015-04-03-T12:07:12+0000 [INFO] Block: 350532 (332.70s)
```

It is inconsistent.  Some blocks parse quickly.  Others don't.

**Comment by deweller** [comment]
My database write-ahead log is quite large (2 GB):

```
-rw-r--r-- 1 xcpd xcpd 615M Apr  3 12:01 counterparty.db
-rw-r--r-- 1 xcpd xcpd  16M Apr  3 12:07 counterparty.db-shm
-rw-r--r-- 1 xcpd xcpd 2.0G Apr  3 12:07 counterparty.db-wal
```

Perhaps this has something to do with it?

**Comment by adamkrellenstein** [comment]
@ouziel-slama: can you look into this?

**Comment by arhuaco** [comment]
Blocks are parsing fast here. But the logs are also big:

    -rw-r--r-- 1 dev  dev  576M Mar 17 23:34 counterpartyd.9.db
    -rw-r--r-- 1 dev  dev   14M Mar 17 23:40 counterpartyd.9.db-shm
    -rw-r--r-- 1 dev  dev  1.8G Mar 17 23:40 counterpartyd.9.db-wal

**Comment by arhuaco** [comment]
Now that the block-chain was parsed, the new blocks are taking > 5 minutes to process. So I confirm I'm seeing something similar.

**Comment by rippler** [comment]
The journal files are big because it's set not to shrink (better for performance). It grows as the DB gets more busy, the size is just telling us that we're handling more transactions than before.

It seems that timing displayed by `counterparty-server` may be misleading. For example, if you're catching up, you'll see very short per-block times. Once you catch up (i.e. you're in sync), then it shows whatever the speed of bitcoin-core is. 
As you can imagine, blocks advance at a rate of 600+ seconds per, so it's no wonder that counterparty-server shows the same. 

I'm still looking but it seems neither the journal size nor the times (when in sync) are concerning.

**Comment by rippler** [comment]
Took another look (I had to wait until there's something resembling a peak)

![time-taken](https://cloud.githubusercontent.com/assets/4045519/7006963/0a88813e-dcb8-11e4-9ca1-fbcefc43a660.png)

* Start Bitcoin Core and let it catch up 
* Then start counterparty-server (no client activity)

Observations:
* Processing times were acceptable - all blocks but one (out of 100+) completed within less than 30 seconds
* One block took 500+ seconds to process and that was the block that Bitcoin Core was processing when counterparty-server was started. Subsequent blocks (as few new were added while counterparty-server was catching up) were also processed quickly. The outlier might have been caused by Bitcoin Core cache or something else.
* I did just one test of this type. More would be necessary but it seems there's no big issue here, unless others are seeing something different than I.

**Comment by robby-dermody** [comment]
I can confirm this is an issue. Causes periodic issues with counterwallet servers as well (i.e. during a reorg or where ever block processing is taking much longer than expected) since counterparty-server will turn away API requests from counterblock when the database is not caught up.

During the time of these long block processings, counterparty-server will max out a CPU core. Given that the behavior appears to be CPU-bound, if I had to guess, I'd guess something with SQLLite... either a missing or unoptimized query, or something else. @rippler tried ussing a VACUUM call, and said that didn't seem to help.

**Comment by rippler** [comment]
I'll just document few other findings:
* SQLite settings used by `counterparty-server` differ between Linux and Windows (but so far we have no indications that this matters because we have reports of slowness on the both OS)
* `counterparty-server` uses `secure_delete` (in SQLite) which is probably not required (no secret info in the DB, and also we don't seem to delete much)
* Indexing is set to automatic and that may be causing extra load on SQLite when many transactions are incoming, but considering that there are relatively few inserts (compared to what SQLite can normally do) it shouldn't be a big factor. If we scheduled SQLite to re-index every few minutes, that would cause regular peaks which would make it even harder to troubleshoot so I didn't try to do this.

**Comment by ouziel-slama** [comment]
@rippler: Thank you for the chart. I think the 500 seconds was not spent for the parsing but for the integrity check. See here: https://github.com/CounterpartyXCP/counterpartyd/blob/master/counterpartylib/lib/blocks.py#L1011

@mtbitcoin, @deweller: Most of the parsing time is spent for RPC calls to Bitcoin Core. I think the problem on some configuration come from here. Especially when another application(s) also makes RPC calls or/and when the Bitcoin Core is on another server. 
Can you please run `counterparty-server` with cProfile and send me the generated file:
```
python -m cProfile -o profile ../../env/bin/counterparty-server --testnet start
```
(depends of you configuration you need to adjust the `counterparty-server` path.

**Comment by mtbitcoin** [comment]
@ouziel-slama I am running this on windows with counterparty-server.exe ? Is there a similar option for this that does not use the python interpreter

**Comment by ouziel-slama** [comment]
@mtbitcoin, try with:
```
python -m cProfile -o profile c:\Python34\Scripts\counterparty-server-script.py --testnet start
```

**Comment by ouziel-slama** [comment]
@mtbitcoin, @deweller : can you please try this branch https://github.com/CounterpartyXCP/counterpartyd/tree/batch
All transactions hex for a block are obtained with only one RPC call. On my configuration the parsing is a little bit faster on testnet, especially for blocks with a lot of transactions.

**Comment by ouziel-slama** [comment]
here the average time to parse one transaction on Mainnet:

without batch: 0.069 s
with batch: 0.010 s

**Comment by adamkrellenstein** [comment]
@ouziel-slama: awesome!

**Comment by adamkrellenstein** [comment]
FYI, merged the batch RPC call patch into `develop`. Does this fix the issue?

**Comment by mtbitcoin** [comment]
To apply the develop, I am extracting my develop branch to c:\counterpartyd
then doing a "python setup.py install" from that folder. Is this the right way ?

**Comment by adamkrellenstein** [comment]
@mtbitcoin, yep

**Comment by deweller** [comment]
Huge improvement so far!
```
2015-04-10-T15:12:02+0000 [INFO] Block: 351537 (8.55s)
2015-04-10-T15:12:06+0000 [INFO] Block: 351538 (3.80s)
2015-04-10-T15:12:11+0000 [INFO] Block: 351539 (5.41s)
2015-04-10-T15:12:12+0000 [INFO] Block: 351540 (1.33s)
```

**Comment by deweller** [comment]
Well, I spoke too soon.  I'm still seeing extraordinary long block times. 

```
2015-04-10-T15:12:02+0000 [INFO] Block: 351537 (8.55s)
2015-04-10-T15:12:06+0000 [INFO] Block: 351538 (3.80s)
2015-04-10-T15:12:11+0000 [INFO] Block: 351539 (5.41s)
2015-04-10-T15:12:12+0000 [INFO] Block: 351540 (1.33s)
2015-04-10-T15:23:35+0000 [INFO] Block: 351541 (682.54s)
2015-04-10-T15:23:38+0000 [INFO] Block: 351542 (2.99s)
2015-04-10-T15:34:57+0000 [INFO] Block: 351543 (678.76s)
```

During these long block parsing times, counterparty-server is taking more than 90% of CPU (according to top) and bitcoind is taking less than 10%.

When not parsing a block, counterparty-server takes about 20%-30% of CPU and bitcoind takes about the same.

**Comment by mtbitcoin** [comment]
I am not seeing the shorter times too

[INFO] Block: 351543 (934.17s)←[0m

**Comment by adamkrellenstein** [comment]
@deweller, @mtbitcoin, try disabling these lines:
https://github.com/CounterpartyXCP/counterpartyd/blob/master/counterpartylib/lib/blocks.py#L1011-L1012

**Comment by mtbitcoin** [comment]
@adamkrellenstein That's not doing it for me (with the latest DEVELOP), I am running into the following with those lines uncommented. 

However uncommenting those lines in MASTER did not throw an error

←[31m[ERROR] Unhandled Exception←[0m
Traceback (most recent call last):
  File "C:\Python34\Scripts\counterparty-server-script.py", line 9, in <module>
    load_entry_point('counterparty-cli==1.1.0', 'console_scripts', 'counterparty-server')(
)
  File "c:\python34\lib\site-packages\counterpartycli\__init__.py", line 16, in server_mai
n
    server.main()
  File "c:\python34\lib\site-packages\counterpartycli\server.py", line 129, in main
    server.start_all(db)
  File "c:\python34\lib\site-packages\counterpartylib\server.py", line 358, in start_all
    blocks.follow(db)
  File "c:\python34\lib\site-packages\counterpartylib\lib\blocks.py", line 986, in follow
    raw_transactions = backend.getrawtransaction_batch(txhash_list)
AttributeError: 'module' object has no attribute 'getrawtransaction_batch'←[0m

**Comment by adamkrellenstein** [comment]
@mtbitcoin, I'm saying run `develop` with those lines commented *out*.

**Comment by mtbitcoin** [comment]
doesn't seem to work for me with those lines commented out in develop. only works on master
** type above, I mean commented out (not uncommented)..

**Comment by mtbitcoin** [comment]
I did see improved times though with the lines commented out.. but its still too early too tell. It could be due to the restart

[INFO] Resuming parsing.←[0m
[INFO] Block: 351545 (9.38s)←[0m
[INFO] Block: 351546 (11.22s)←[0m

**Comment by deweller** [comment]
For comparison (running develop *without* any lines commented out):
```
2015-04-10-T16:05:27+0000 [INFO] Block: 351545 (225.82s)
2015-04-10-T16:10:49+0000 [INFO] Block: 351546 (240.82s)
2015-04-10-T16:28:51+0000 [INFO] Block: 351547 (290.29s)
```

**Comment by mtbitcoin** [comment]
Looks promising so far

Server1 (with lines commented out)
[INFO] Send: 3.69566659 FLDC from 15RAB9euaa2SPyUMUWJo9v13LMtkbnKUCW to ...]←[0m
[INFO] Block: 351547 (19.42s)←[0m

Server2 (WITHOUT lines commented out)
[INFO] Send: 3.69566659 FLDC from 15RAB9euaa2SPyUMUWJo9v13LMtkbnKUCW to ...]←[0m
[INFO] Block: 351547 (434.05s)←[0m

From what i observed both has finished parsing all of the transactions.  Meaning the last transactions lines above "[INFO] Send: 3.69566659 FLDC from 15RAB9euaa2SPyUMUWJo9v13LMtkbnKUCW to ...]←[0m" appeared about the same time but there was no block completion time for Server2 until many minutes later.  Could these be due to the additional process in "check.asset_conservation?"

Counterparty-server was restarted at the same time on both servers

**Comment by adamkrellenstein** [comment]
So I think the way to speed up `check.asset_conservation` is to solve this issue:
https://github.com/CounterpartyXCP/counterpartyd/issues/347

@ouziel-slama, you want to tackle that?

**Comment by mtbitcoin** [comment]
will leaving check.asset_conservation have any repurcussions for now? I am running this on a live server

**Comment by adamkrellenstein** [comment]
@mtbitcoin, it's safe to leave it commented out.

**Comment by mtbitcoin** [comment]
I'll do a repost of the block if there is still slowdown overnite

**Comment by adamkrellenstein** [comment]
It seems to me that the whole problem is very likely `check.asset_conservation`. That's the only thing that explains the CPU usage.

Note: this isn't a problem only with v9.51.0. There must've just been a lot of recent asset issuances.

**Comment by ouziel-slama** [comment]
@adamkrellenstein: sure!

**Comment by deweller** [comment]
Running master with the `check.asset_conservation` lines commented out:
```
2015-04-10-T16:45:56+0000 [INFO] Block: 351548 (6.50s)
2015-04-10-T16:49:21+0000 [INFO] Block: 351549 (1.71s)
2015-04-10-T16:49:56+0000 [INFO] Block: 351550 (0.84s)
2015-04-10-T17:05:04+0000 [INFO] Block: 351551 (5.04s)
2015-04-10-T17:13:43+0000 [INFO] Block: 351552 (3.69s)
2015-04-10-T17:28:17+0000 [INFO] Block: 351553 (4.66s)
2015-04-10-T17:33:42+0000 [INFO] Block: 351554 (1.34s)
2015-04-10-T17:34:23+0000 [INFO] Block: 351555 (0.29s)
2015-04-10-T17:34:59+0000 [INFO] Block: 351556 (2.29s)
```
I think you have correctly identified the source of the slowdown.  And, yes, there have been a lot of assets registered recently with the launch of CoinDaddy.

**Comment by adamkrellenstein** [comment]
Closing. Actual issue is https://github.com/CounterpartyXCP/counterpartyd/issues/752.

**Comment by adamkrellenstein** [comment]
@mtbitcoin, @deweller, @robby-dermody, the fix was just merged into `develop`: #754
